zsh plugins: use omz-compatible paths

LuigiPiucco opened the pull request:
Note: This is my first pull request, I apologize in advance for
anything bad or non-standard I may have done. Point me in the right
direction and I'll try to correct it.

###### Motivation for this change

This allows for easier configuration of zsh through NixOS configuration.

The `programs.zsh.ohMyZsh.customPkgs` option expects that arguments
passed to it will have a zsh plugin/theme/completion in a path like
`$out/share/zsh/{plugins,themes,completions}`, and copies whatever is
inside those to `$ZSH_CUSTOM`.

However, some packages actually unpacked to `$out/share`, without the
proper folder structure, so they never got copied to `$ZSH_CUSTOM` and
caused "not found" errors on zsh startup.

This fixes that, and maintains compatibility with other packages
that might still expect a flatted structure. This is done though a
symlink.

Previously, one could only (at least, it was the only thing that worked
for me) add things like autosuggestions and syntax highlighting though
options like `programs.zsh.autosuggestions.enable` and 
`programs.zsh.syntaxHighlighting.enable`. Those are still supported,
but now adding the appropriate package to
`programs.zsh.ohMyZsh.customPkgs` and listing its name in 
`programs.zsh.ohMyZsh.plugins` works too. 

###### Things done

Some packages which unpacked to an improper location where refactored
and symlinked to their old path. I checked all zsh packages, the
unchanged ones seemed to already be working though some are untested. I
can try and test them if requested. 

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

LuigiPiucco committed:
zsh plugins: use omz-compatible paths

The `programs.zsh.ohMyZsh.customPkgs` option expects that arguments
passed to it will have a zsh plugin/theme/completion in a path like
`$out/share/zsh/{plugins,themes,completions}`, and copies whatever is
inside those to `$ZSH_CUSTOM`.

However, some packages actually unpacked to `$out/share`, without the
proper folder structure, so they never got copied to `$ZSH_CUSTOM` and
caused "not found" errors on zsh startup.

This commit fixes that, and maintains compatibility with other packages
that might still expect a flattened structure. This is done though a
symlink.

Janik-Haag commented:
Hi, I'm so sorry no one reviewed this especially since this was your first pr. I sadly don't know zsh well enough to reviews this. Are you still interested in getting this merged, if you are just ping me and I will ask in the matrix if some one could review this.

LuigiPiucco commented:
@Janik-Haag

So, 4 years have passed, and I've come to get closure on this. I rebased and updated it. Note that I did not test the new push, as I don't use zsh anymore. I can if requested, but I remember having made sure it worked when I originally sent it, and there was only a small conflict and a misspelling.

JohnRTitor commented on pkgs/by-name/zs/zsh-bd/package.nix:
```diff
@@ -19,10 +19,16 @@ stdenv.mkDerivation {
+    install -D bd.zsh \
+      $out/share/plugins/zsh-bd/bd.zsh
+    ln -s $out/share/plugins/zsh-bd/bd.zsh \
+      $out/share/plugins/zsh-bd/bd.plugin.zsh
```
Build failure was due to a typo, fixed.
